src/Build/BuildCheck/API/BuildCheckResult.cs

```diff
@@ -32,9 +28,9 @@ public BuildCheckResult(CheckRule checkConfig, IMSBuildElementLocation location,
     internal BuildEventArgs ToEventArgs(CheckResultSeverity severity)
         => severity switch
         {
-            CheckResultSeverity.Suggestion => new BuildCheckResultMessage(this),
-            CheckResultSeverity.Warning => new BuildCheckResultWarning(this, CheckRule.Id),
-            CheckResultSeverity.Error => new BuildCheckResultError(this, CheckRule.Id),
+            CheckResultSeverity.Suggestion => new BuildCheckResultMessage(FormatMessage()),
+            CheckResultSeverity.Warning => new BuildCheckResultWarning(FormatMessage(), CheckRule.Id),
+            CheckResultSeverity.Error => new BuildCheckResultError(FormatMessage(), CheckRule.Id),
```

Comment:
I'd suggest to create a separate PR for this.

What we should really do is to keep passing the `IBuildCheckResult` here, and exposte the info required downstream via that interface and then properly consume it downstream. E.g. the `BuildCheckResultError` ctor doesn't properly populate location information into it's base constructor.

Reply:
I created https://github.com/dotnet/msbuild/issues/10529